### ISSUE TITLE
K8s ingress sample: Missing Cluster Id exception

### ISSUE THREAD

**Issue by jtkech** [microsoft/reverse-proxy#1092]
I started to learn `Yarp` to see how we could use it for `OrchardCore`. And i'm a beginner on k8s, so i followed the `kubernetes-ingress.md` article.

### Describe the bug

One step of the article is `kubectl apply -f ingress-definition.yaml` but this yaml file doesn't exist, so I replaced it with `-f ingress-controller.yaml`. Then another step is `kubectl apply -f .\backend.yaml` but this yaml file doesn't exist, so I created it to define a backend deployment and service.

Then the `yarp-proxy` was failing on `The proxy config is invalid. (Missing Cluster Id.)`. In the below logs we can see that in the `Cluster` property of the message an `Id` is passed, but it was expecting a `ClusterId`. So, my workaround was to rename in `ClusterConfig` the `ClusterId` prop to `Id`, and rebuild the image, then it was working.

 https://github.com/microsoft/reverse-proxy/blob/f75df4d7ec6fb23c8f7af53e853e6fe9ed02cf6e/src/ReverseProxy/Configuration/ClusterConfig.cs#L20

### To Reproduce

Follow the `kubernetes-ingress.md` article.

Part of the logs in the `yarp-proxy` pod.

    [14:09:27 INF] Received Update for default:minimal-ingress
    [14:09:27 INF] {"MessageType":"Update","Key":"default:minimal-ingress",
      "Routes":[{"RouteId":"/","Match":{"Methods":null,"Hosts":null,"Path":"/{**catch-all}","Headers":null},"Order":null,"ClusterId":"10.1.0.108","AuthorizationPolicy":null,"CorsPolicy":null,"Metadata":null,"Transforms":null}],
      "Cluster":[{"Id":"10.1.0.108","LoadBalancingPolicy":null,"SessionAffinity":null,"HealthCheck":null,"HttpClient":null,"HttpRequest":null,"Destinations":{"http://10.1.0.108:8080":{"Address":"http://10.1.0.108:8080","Health":null,"Metadata":null}},"Metadata":null}]}
    [14:09:27 INF] Update
    [14:09:27 ERR] Failed to apply the new config.
    System.AggregateException: The proxy config is invalid. (Missing Cluster Id.)
    ---> System.ArgumentException: Missing Cluster Id.
    --- End of inner exception stack trace ---
    at Yarp.ReverseProxy.Management.ProxyConfigManager.ApplyConfigAsync(IProxyConfig config) in /src/src/ReverseProxy/Management/ProxyConfigManager.cs:line 217
    at Yarp.ReverseProxy.Management.ProxyConfigManager.ReloadConfigAsync() in /src/src/ReverseProxy/Management/ProxyConfigManager.cs:line 187

### Further technical details

- I cloned the repository 2 days ago.
- The platform: Windows

**Comment by karelz** [comment]
Looks like you found some bugs in the sample.

The last one (difference in `ClusterId` and `Id`) -- this is intentional breaking change introduced in PR #963 (will be part of Preview 12).
Looks like we need to update also the Kubernetes Docker now to reflect the config changes? @Tratcher any insights here?

**Comment by Tratcher** [comment]
ingress-controller.yaml references an old sample image from one of our devs that we need to replace with something more official. We haven't set up the process to build or maintain those images yet.
https://github.com/microsoft/reverse-proxy/blob/f75df4d7ec6fb23c8f7af53e853e6fe9ed02cf6e/samples/KuberenetesIngress.Sample/Ingress/ingress-controller.yaml#L128
For now it would be better to build https://github.com/microsoft/reverse-proxy/tree/main/src/ReverseProxy.Kubernetes.Controller locally.

As for backend.yml, it looks like that became https://github.com/microsoft/reverse-proxy/blob/main/samples/KuberenetesIngress.Sample/backend/ingress-sample.yaml

A PR to clean up the readme would be welcome.

**Comment by jtkech** [comment]
Thanks for the info

- First, I was still using the existing `ingress-sample.yaml` that defines the minimal ingress and that point to the backend service, so I really needed to create a new `backend.yaml` to define the backend deployment and service.

- So I built the yarp controller image locally and pushed it on my registry. Note: I needed to update some paths in the existing `Dockerfile.Multifile` to be able to use it. With this updated image, in the logs there is no more exception and i can see the requests, but there is no update message from the minimal ingress, so there is no matched endpoint.

- So for testing i re-applied my workaround by renaming `ClusterId` to `Id`, but it was still not working because there is still no update from the minimal ingress, so still no route to match.

- Finally, because my workaround was re-applied, i just re-used the old `jkotalik/yarp-controller:0.0.0-19c9e0d733` image, and then i could see in the logs `Received Update for default:minimal-ingress`, and it was working.

No problem to update the article if the above can be clarified.
Nothing urgent for me, just to share my results if that may help.

**Comment by Tratcher** [comment]
Thanks @jtkech. We're focused on finishing up the core library right now so we'll come back to K8 in a little while. Any cleanup contributions you want to make in the meantime would be appreciated.

**Comment by jtkech** [comment]
@Tratcher Okay no problem I can wait, I will do some first suggestions through a PR.

**Comment by jtkech** [comment]
@Tratcher  Just for info.

- I updated the doc and did some little cleanups.

- I could get it fully working, just needed in the `DispatchController` to replace:

      return new Task<IActionResult>(() => new DispatchActionResult(...));

    With:

      return Task.FromResult<IActionResult>(new DispatchActionResult(...));

I will do a PR soon, for now, because I only cloned the repo and i don't have write accesses, I can't push my changes from my local branch. So, I think i need to fork the repo to my own repo first, will see tomorrow.
